File: src/topo_shadow_box/core/osm.py
"""OpenStreetMap feature fetching via Overpass API."""

import asyncio
import logging

import httpx

from .models import OsmFeatureSet
from topo_shadow_box.models import RoadFeature, WaterFeature, BuildingFeature, Coordinate

logger = logging.getLogger(__name__)

OVERPASS_SERVERS = [
    "https://overpass.kumi.systems/api/interpreter",
    "https://overpass-api.de/api/interpreter",
    "https://overpass.openstreetmap.ru/api/interpreter",
]


async def _query_overpass(query: str) -> list[dict]:
    """Execute an Overpass API query with server fallback."""
    async with httpx.AsyncClient(timeout=45.0, headers={"User-Agent": "topo-shadow-box/1.0"}) as client:
        for server in OVERPASS_SERVERS:
            try:
                response = await client.post(server, data={"data": query})
                response.raise_for_status()
                data = response.json()
                return data.get("elements", [])
            except httpx.TimeoutException as exc:
                logger.warning("Overpass server %s timed out: %s", server, exc)
                continue
            except httpx.HTTPStatusError as exc:
                logger.warning(
                    "Overpass server %s returned HTTP %s", server, exc.response.status_code
                )
                continue
            except Exception as exc:
                logger.warning("Overpass server %s failed: %s", server, exc)
                continue
    logger.warning("All Overpass servers failed for query")
    return []


def _parse_way_coords(element: dict) -> list[dict]:
    """Extract coordinate list from a way or relation element."""
    coords = []
    elem_type = element.get("type")

    if elem_type == "way":
        for pt in element.get("geometry", []):
            coords.append({"lat": pt["lat"], "lon": pt["lon"]})
    elif elem_type == "relation":
        for member in element.get("members", []):
            if member.get("role") == "outer" and member.get("type") == "way":
                for pt in member.get("geometry", []):
                    coords.append({"lat": pt["lat"], "lon": pt["lon"]})
                break  # Use first outer ring
    return coords


def _parse_features(elements: list[dict], feature_type: str) -> list:
    """Parse Overpass elements into typed feature models."""
    features = []
    for elem in elements:
        raw_coords = _parse_way_coords(elem)
        if len(raw_coords) < 2:
            continue
        coords = [Coordinate(lat=c["lat"], lon=c["lon"]) for c in raw_coords]
        tags = elem.get("tags", {})
        name = tags.get("name", f"{feature_type}_{elem.get('id')}")

        if feature_type == "road":
            if len(coords) < 2:
                continue
            features.append(RoadFeature(
                id=elem.get("id", 0),
                coordinates=coords,
                tags=tags,
                name=name,
                road_type=tags.get("highway", "road"),
            ))
        elif feature_type == "water":
            if len(coords) < 3:
                continue
            features.append(WaterFeature(
                id=elem.get("id", 0),
                coordinates=coords,
                tags=tags,
                name=name,
            ))
        elif feature_type == "building":
            if len(coords) < 3:
                continue
            height = 10.0
            if "height" in tags:
                try:
                    height = float(tags["height"].replace("m", "").strip())
                except ValueError:
                    pass
            elif "building:levels" in tags:
                try:
                    height = float(tags["building:levels"]) * 3.0
                except ValueError:
                    pass
            height = max(height, 0.1)  # ensure gt=0 for BuildingFeature
            features.append(BuildingFeature(
                id=elem.get("id", 0),
                coordinates=coords,
                tags=tags,
                name=name,
                height=height,
            ))
    return features


async def fetch_osm_features(
    north: float, south: float, east: float, west: float,
    feature_types: list[str],
) -> OsmFeatureSet:
    """Fetch OSM features for a bounding box.

    Args:
        feature_types: List of types: 'roads', 'water', 'buildings'

    Returns:
        OsmFeatureSet with typed feature lists.
    """
    bbox = f"{south},{west},{north},{east}"

    road_elements = []
    water_elements = []
    building_elements = []

    queries = {}
    if "roads" in feature_types:
        queries["roads"] = (
            f'[out:json][timeout:30];way["highway"]({bbox});out body geom;',
            "road",
        )
    if "water" in feature_types:
        queries["water"] = (
            f'[out:json][timeout:30];(way["natural"="water"]({bbox});'
            f'way["waterway"]({bbox});'
            f'relation["natural"="water"]({bbox}););out body geom;',
            "water",
        )
    if "buildings" in feature_types:
        queries["buildings"] = (
            f'[out:json][timeout:30];way["building"]({bbox});out body geom;',
            "building",
        )

    queries_list = list(queries.items())
    for i, (feat_name, (query, parse_type)) in enumerate(queries_list):
        if i > 0:
            await asyncio.sleep(1.0)  # OSM rate limit: 1 req/sec
        try:
            elements = await _query_overpass(query)
        except Exception:
            elements = []
        if feat_name == "roads":
            road_elements = elements
        elif feat_name == "water":
            water_elements = elements
        elif feat_name == "buildings":
            building_elements = elements

    roads = _parse_features(road_elements, "road")[:200]
    water = _parse_features(water_elements, "water")[:50]
    buildings = _parse_features(building_elements, "building")[:150]
    return OsmFeatureSet(roads=roads, water=water, buildings=buildings)

File: go/internal/core/osm.go
package core

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"

	"github.com/huslage/topo-shadow-box/internal/session"
)

// OverpassClient sends Overpass API requests (POST-based).
type OverpassClient interface {
	Post(url, contentType string, body io.Reader) (*http.Response, error)
}

var overpassServers = []string{
	"https://overpass-api.de/api/interpreter",
	"https://overpass.kumi.systems/api/interpreter",
}

// FetchOSMFeatures fetches OSM roads, water, and/or buildings for the given
// bounding box via the Overpass API. Pass nil for client to use http.DefaultClient.
// featureTypes is a subset of ["roads", "water", "buildings"].
func FetchOSMFeatures(ctx context.Context, client OverpassClient, north, south, east, west float64, featureTypes []string) (*session.OsmFeatureSet, error) {
	if client == nil {
		client = http.DefaultClient
	}
	result := &session.OsmFeatureSet{}
	wantRoads := contains(featureTypes, "roads")
	wantWater := contains(featureTypes, "water")
	wantBuildings := contains(featureTypes, "buildings")

	bbox := fmt.Sprintf("%.6f,%.6f,%.6f,%.6f", south, west, north, east)

	if wantRoads {
		query := fmt.Sprintf(`[out:json][timeout:25];(way["highway"~"^(motorway|trunk|primary|secondary|tertiary|residential|service|unclassified)$"](%s););out geom;`, bbox)
		elements, err := runOverpassQuery(ctx, client, query)
		if err == nil {
			for _, el := range elements {
				if el.Type != "way" {
					continue
				}
				road := session.RoadFeature{
					ID:       int(el.ID),
					Name:     el.Tags["name"],
					RoadType: el.Tags["highway"],
					Tags:     el.Tags,
				}
				for _, g := range el.Geometry {
					road.Coordinates = append(road.Coordinates, session.Coordinate{Lat: g.Lat, Lon: g.Lon})
				}
				if len(road.Coordinates) >= 2 {
					result.Roads = append(result.Roads, road)
				}
			}
			if len(result.Roads) > 200 {
				result.Roads = result.Roads[:200]
			}
		}
	}

	if wantWater {
		query := fmt.Sprintf(`[out:json][timeout:25];(way["natural"~"^(water|coastline)$"](%s);way["waterway"](%s);relation["natural"="water"](%s););out geom;`, bbox, bbox, bbox)
		elements, err := runOverpassQuery(ctx, client, query)
		if err == nil {
			for _, el := range elements {
				if el.Type != "way" {
					continue
				}
				water := session.WaterFeature{
					ID:   int(el.ID),
					Name: el.Tags["name"],
					Tags: el.Tags,
				}
				for _, g := range el.Geometry {
					water.Coordinates = append(water.Coordinates, session.Coordinate{Lat: g.Lat, Lon: g.Lon})
				}
				if len(water.Coordinates) >= 2 {
					result.Water = append(result.Water, water)
				}
			}
			if len(result.Water) > 50 {
				result.Water = result.Water[:50]
			}
		}
	}

	if wantBuildings {
		query := fmt.Sprintf(`[out:json][timeout:25];(way["building"](%s););out geom;`, bbox)
		elements, err := runOverpassQuery(ctx, client, query)
		if err == nil {
			for _, el := range elements {
				if el.Type != "way" {
					continue
				}
				b := session.BuildingFeature{
					ID:   int(el.ID),
					Name: el.Tags["name"],
					Tags: el.Tags,
				}
				// Height parsing: prefer explicit "height" tag, fall back to building:levels
				if h, ok := el.Tags["height"]; ok {
					h = strings.TrimSuffix(strings.TrimSpace(h), "m")
					if v, err2 := strconv.ParseFloat(strings.TrimSpace(h), 64); err2 == nil {
						b.Height = v
					}
				} else if lvl, ok := el.Tags["building:levels"]; ok {
					if v, err2 := strconv.ParseFloat(strings.TrimSpace(lvl), 64); err2 == nil {
						b.Height = v * 3.0
					}
				}
				if b.Height <= 0 {
					b.Height = 6.0 // default: 2 floors at 3m each
				}
				for _, g := range el.Geometry {
					b.Coordinates = append(b.Coordinates, session.Coordinate{Lat: g.Lat, Lon: g.Lon})
				}
				if len(b.Coordinates) >= 3 {
					result.Buildings = append(result.Buildings, b)
				}
			}
			if len(result.Buildings) > 150 {
				result.Buildings = result.Buildings[:150]
			}
		}
	}

	return result, nil
}

type overpassElement struct {
	Type     string            `json:"type"`
	ID       int64             `json:"id"`
	Tags     map[string]string `json:"tags"`
	Geometry []struct {
		Lat float64 `json:"lat"`
		Lon float64 `json:"lon"`
	} `json:"geometry"`
}

type overpassResponse struct {
	Elements []overpassElement `json:"elements"`
}

func runOverpassQuery(ctx context.Context, client OverpassClient, query string) ([]overpassElement, error) {
	var lastErr error
	for _, server := range overpassServers {
		resp, err := client.Post(server, "application/x-www-form-urlencoded", strings.NewReader("data="+query))
		if err != nil {
			lastErr = err
			continue
		}
		defer resp.Body.Close()
		if resp.StatusCode != 200 {
			lastErr = fmt.Errorf("overpass returned %d", resp.StatusCode)
			continue
		}
		data, err := io.ReadAll(resp.Body)
		if err != nil {
			lastErr = err
			continue
		}
		var result overpassResponse
		if err := json.Unmarshal(data, &result); err != nil {
			lastErr = err
			continue
		}
		return result.Elements, nil
	}
	return nil, fmt.Errorf("all overpass servers failed: %w", lastErr)
}

func contains(s []string, v string) bool {
	for _, x := range s {
		if x == v {
			return true
		}
	}
	return false
}
